### PR TITLE
BUG: Fix difference functionality for Period Indexes

### LINKED ISSUE #58971: BUG: `Index.difference` is returning too many values
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

index1 = pd.period_range('2022-01-01', periods=5, freq='M')
index2 = pd.Index(['2022-02', '2022-03'])
result = index1.difference(index2)
print(result)
```

### Issue Description

The example creates `index1` with 5 months, and `index2` with two months. The difference should be an index with three months. However, I'm getting the following:

`PeriodIndex(['2022-01', '2022-02', '2022-03', '2022-04', '2022-05'], dtype='period[M]')`

Looks like this issue was introduced by https://github.com/pandas-dev/pandas/pull/55108.

### Expected Behavior

The example should print the following:

`Index(['2022-02', '2022-03'], dtype='object')`

Pandas 2.1.4 gives the expected behavior.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit                : 6dbeeb4009bbfac5ea1ae2111346f5e9f05b81f4
python                : 3.10.8.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 6.5.0-35-generic
Version               : #35-Ubuntu SMP PREEMPT_DYNAMIC Fri Apr 26 11:23:57 UTC 2024
machine               : x86_64
processor             : 
byteorder             : little
LC_ALL                : None
LANG                  : C.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.0rc0+28.g6dbeeb4009
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 63.2.0
pip                   : 24.0
Cython                : 3.0.10
pytest                : 8.2.0
hypothesis            : 6.100.2
sphinx                : 7.3.7
blosc                 : None
feather               : None
xlsxwriter            : 3.2.0
lxml.etree            : 5.2.1
html5lib              : 1.1
pymysql               : 1.4.6
psycopg2              : 2.9.9
jinja2                : 3.1.3
IPython               : 8.24.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : 1.3.8
fastparquet           : 2024.2.0
fsspec                : 2024.3.1
gcsfs                 : 2024.3.1
matplotlib            : 3.8.4
numba                 : 0.59.1
numexpr               : 2.10.0
odfpy                 : None
openpyxl              : 3.1.2
pyarrow               : 16.0.0
pyreadstat            : 1.2.7
python-calamine       : None
pyxlsb                : 1.0.10
s3fs                  : 2024.3.1
scipy                 : 1.13.0
sqlalchemy            : 2.0.29
tables                : 3.9.2
tabulate              : 0.9.0
xarray                : 2024.3.0
xlrd                  : 2.0.1
zstandard             : 0.22.0
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None
</details>

### PR DESCRIPTION
**Pull request by TiffL**
- [x] closes #58971
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by TiffL** [comment]
Hi @phofl and @MarcoGorelli! Contributing as part of Bloomberg's open source sustainability series, would you be able to help review this PR? Thank you!

**Review by WillAyd — APPROVED** [review]
lgtm @mroeschke @jbrockmendel any concerns?

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -514,6 +514,15 @@ def shift(self, periods: int = 1, freq=None) -> Self:
             )
         return self + periods
 
+    def _difference(self, other, sort=None) -> PeriodIndex:
+        if isinstance(other, Index) and other.inferred_type == "string":
+            try:
+                other = other.astype(self.dtype)
```
Comment (jbrockmendel) [inline pandas/core/indexes/period.py:520]: I don’t think we do this type of casting in general. In fact, I’d expect the private _difference to only be reached with matching dtypes

File: pandas/tests/indexes/period/test_setops.py [hunk 2]
```diff
@@ -314,6 +314,19 @@ def test_difference(self, sort):
...
+        # GH58971
+        index = period_range("2022-01-01", periods=5, freq="M")
+        other = pd.Index(["2022-02", "2022-03"])
+
+        idx_diff = index.difference(other, sort)
+        expected = PeriodIndex(["2022-01", "2022-04", "2022-05"], freq="M")
+        tm.assert_index_equal(idx_diff, expected)
+
+        idx_diff = other.difference(index, sort)
+        expected = pd.Index([])
```
Comment (jbrockmendel) [inline pandas/tests/indexes/period/test_setops.py:327]: No, the difference should be the entire ‘other’

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**